google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/BuildDeploymentSourceType.java

```diff
@@ -0,0 +1,124 @@
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.configuration.deployment.ModuleDeploymentSource;
+import com.intellij.remoteServer.impl.configuration.deployment.DeployToServerRunConfiguration;
+import com.intellij.remoteServer.impl.configuration.deployment.ModuleDeploymentSourceType;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.List;
+
+import javax.swing.JComponent;
+
+/**
+ * Base class for build system (e.g. maven or gradle) deployment source types.
+ */
+public abstract class BuildDeploymentSourceType extends ModuleDeploymentSourceType {
+
+  @Override
+  public void setBuildBeforeRunTask(
+      @NotNull RunConfiguration configuration,
+      @NotNull ModuleDeploymentSource source) {
+
+    Module module = source.getModule();
+
+    if(module == null) {
+      return;
+    }
+
+    setConfiguration(configuration);
+
+    RunManagerEx runManager = RunManagerEx.getInstanceEx(configuration.getProject());
+    final List<? extends BeforeRunTask> buildTasks = getBuildTasks(runManager, configuration);
+
+    if (hasBuildTaskForModule(buildTasks, module)) {
+      for (BeforeRunTask task : buildTasks) {
+        task.setEnabled(true);
+      }
+    } else {
+      BeforeRunTask buildTask = createBuildTask(module);
+      List<BeforeRunTask> tasks = runManager.getBeforeRunTasks(configuration);
+      tasks.add(buildTask);
+      runManager.setBeforeRunTasks(configuration, tasks, true);
+    }
+  }
+
+  @Override
+  public void updateBuildBeforeRunOption(
+      @NotNull JComponent runConfigurationEditorComponent,
+      @NotNull Project project,
+      @NotNull ModuleDeploymentSource source,
+      boolean select) {
+
+    final DataContext dataContext =
+        DataManager.getInstance().getDataContext(runConfigurationEditorComponent);
+    final ConfigurationSettingsEditorWrapper editor =
+        ConfigurationSettingsEditorWrapper.CONFIGURATION_EDITOR_KEY.getData(dataContext);
+
+    Module module = source.getModule();
+
+    if(module != null && editor != null) {
+      BeforeRunTask buildTask = createBuildTask(module);
+
+      if(buildTask != null) {
+        List<BeforeRunTask> beforeRunTasks = editor.getStepsBeforeLaunch();
+        if (select && !hasBuildTaskForModule(beforeRunTasks, module)) {
+          editor.addBeforeLaunchStep(buildTask);
+        }
+      }
+    }
+  }
+
+  @NotNull
+  protected abstract List<? extends BeforeRunTask> getBuildTasks(
+      RunManagerEx runManager,
+      RunConfiguration configuration);
+
+  @Nullable
+  protected abstract BeforeRunTask createBuildTask(Module module);
+
+  protected abstract boolean hasBuildTaskForModule(
+      Collection<? extends BeforeRunTask> beforeRunTasks, Module module);
+
+  /**
+   * Manually set the deployment configuration so that its available immediately in the
```

Comment:
It does break other ModuleDeploymentSourceTypes. It also breaks ArtifactDeploymentSourceTypes for built in IJ artifacts. Attempting to deploy to a pre-selected IJ artifact will throw an NPE unless a similar fix is in place (or the user edits the UI). Also interesting to note that the existing App Engine plugin has this same issue